zulu: add version check in passthru.tests

msgilligan opened the pull request:
This basic check just launches `java -version` and makes sure it returns some expected output.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Infinidoge reviewed (approved):
Nit: the commit message should probably be `zulu: add installCheckPhase`

Besides that, looks good to me

msgilligan committed:
zulu: add installCheckPhase

This basic check just launches `java -version` and makes sure
it returns some expected output.

dtomvan commented on pkgs/development/compilers/zulu/common.nix:
```diff
@@ -141,6 +141,18 @@ let
+    installCheckPhase = ''
+      echo "Checking 'java -version' output contains ${dist.jdkVersion}, 'OpenJDK', and 'Zulu'"
+      output=`$out/bin/java -version 2>&1`
+      # Our Zulu 23 package is a special case (and should be deleted as it is End-of-Life)
```
Is there a PR for that or shall we make one?

msgilligan replied:
PR #421587, but see also PR #403247

dtomvan commented on pkgs/development/compilers/zulu/common.nix:
```diff
@@ -141,6 +141,18 @@ let
         fi
       '';
 
+    doInstallCheck = true;
+    installCheckPhase = ''
+      echo "Checking 'java -version' output contains ${dist.jdkVersion}, 'OpenJDK', and 'Zulu'"
+      output=`$out/bin/java -version 2>&1`
+      # Our Zulu 23 package is a special case (and should be deleted as it is End-of-Life)
+      expected_version=$(if [[ "${dist.jdkVersion}" == 23* ]]; then echo 23; else echo "${dist.jdkVersion}"; fi)
+      echo "Output is: $output"
+      echo "$output" | grep -q "$expected_version"
+      echo "$output" | grep -q "OpenJDK"
+      echo "$output" | grep -q "Zulu"
+    '';
+
```
Nit: you don't need to hand-roll your own bash script here. See https://github.com/NixOS/nixpkgs/blob/master/doc/build-helpers/testers.chapter.md:

```suggestion
  passthru.tests.version = testers.testVersion {
    package = jdk;
    command = "java -version";
    version = ''openjdk version "${if lib.versions.major version == "23" then "23" else version}"'';
  };
```

This does have some subtle differences though:
- It won't look for `OpenJDK` or `Zulu`
- It doesn't get run unless done so explicitly with `nix-build -A zulu.passthru.tests`

msgilligan replied:
Oh, that's nice!

Maybe these tests should be `passthru` tests anyway?

Maybe `testVersion` could be extended to also take an optional list of additional keywords (or patterns) that must be found?

msgilligan replied:
> it's not that thorough of a functionality test anyways.

Fair enough. We don't _need_ it. I thought it was nice to check that it is actually the distribution of OpenJDK it claims to be. But definitely not necessary.

msgilligan replied:
@dtomvan I merged your commit, but it was failing checks, so I added another commit that seems to be working. Apparently the issue with `zulu23` wasn't limited to `zulu23` but can happen on any "nn.0.0" version because the ".0.0" part is not displayed in the version output.

It took me a couple of tries to get this right, but I think the current force-push should pass the tests.

msgilligan committed:
Use pass thru.tests.version not custom installCheckPhase script

Co-authored-by: Tom van Dijk <18gatenmaker6@gmail.com>

msgilligan committed:
zulu: Fix passthru.tests

* Eliminate duplicate passthru definition by moving "tests.version"
* Only compare major version (because the ".0.0" can be truncated
* Special case major version 8 to "1.8" for the comparison

Aleksanaa commented:
Sorry, didn't edit title and squash merge
